### ISSUE TITLE
Utilize Lparallel

### ISSUE THREAD

**Issue by jmercouris** [atlas-engineer/nyxt#1683]
If we utilize lparallel (particularly in prompter), I think we'll see great performance increases by the usage of paralell mapcar etc.

https://lparallel.org/cognates/

@Ambrevar

**Comment by Ambrevar** [comment]
lparallel does not have the SELECT statement like Calispel does.
(It also has an uncommon definition of "channels".)

If what you have in mind is to call object-properties in parallel, in
would only defer the issue: you'll divide the rendering time by the
number of cores you have.  Not much of an improvement in my opinion (at
the cost of huge CPU spikes :p).

Or did you have something else in mind?

Optimization ideas for the prompt buffer (it's been on the road map
since the beginning, just need to find time to work on it):

- Lazy evaluation of object-properties
- Caching / memoization of object-properties.

Combined, you would essentially reduce the object-properties computation
to practically nothing.

**Comment by jmercouris** [comment]
I have no desire to replace calispel, I just want to use some of lparallel processing facilities to increase the performance of some of our mapcar operations in the codebase.

**Comment by Ambrevar** [comment]
We could implement pmap for Calispel if we want to save a dependency,
it's trivial.

That said, I believe it's not a very promising strategy in the short
term.  I'd like to focus on the aforementioned ideas, they seem much
more promising! :)

**Comment by jmercouris** [comment]
I think the cognates are very useful and could result in a huge speedup in many situations. The trend these days is largely towards parallelization of workloads. Many people have 16, 32 core processors. Utilizing lparallel cognate allow us to easily take advantage of those extra cores. There is no reason that something like mapcar needs to be single threaded. I believe the only reason it is is because concurrency was not included in the original CL spec.

**Comment by Ambrevar** [comment]
It needs to be used sparingly.  Replacing all mapcar with pmap won't do
when the cost of firing up threads outweighs the processing of the
list.

For short lists (probably all of them) and lightweight operations, pmap
may be slower.  Benchmarking is necessary to avoid premature optimization.

Most operations performed by Nyxt are very lightweight.

**Comment by jmercouris** [comment]
It doesn't need to fire up threads, it maintains some worker threads in the kernel. I'm not suggesting replacing all instances!

**Comment by jmercouris** [comment]
It reuses them :-D

**Comment by Ambrevar** [comment]
> It doesn't need to fire up threads, it maintains some worker threads
> in the kernel.

Fair enough.  Then I'm not sure what happens when you run multiple pmap
concurrently.

Lparallel and its kernel model is meant for a specific use case which
does not match Nyxt very well so far.

> I'm not suggesting replacing all instances!

Do you have a specific code location in mind?

Generally speaking, Nyxt does not do much heavy processing at all, which
is why I believe thinking parallelization is a bit premature at the
moment.

**Comment by jmercouris** [comment]
> It doesn't need to fire up threads, it maintains some worker threads in the kernel.
> Fair enough. Then I'm not sure what happens when you run multiple pmap concurrently. Lparallel and its kernel model is meant for a specific use case which does not match Nyxt very well so far.

You cannot run multiple pmap concurrently.

Consider:

```
(pmap ...)
(pmap ...)
```

just because pmap is a parallel operation does not mean it is not a blocking operation! The only way to run multiple pmaps simultaneously would be if they originated on separate threads. In that case, they would share the workers until the task is completed, no slower than a normal mapcar.

> I'm not suggesting replacing all instances!
> Do you have a specific code location in mind? Generally speaking, Nyxt does not do much heavy processing at all, which is why I believe thinking parallelization is a bit premature at the moment.

I don't have any specific things in mind, I would like to add it to our toolbelt though so that we can use it when it is applicable.

**Comment by jmercouris** [comment]
I say it is no slower than a normal mapcar because in both cases two mapcars running on two threads will also be competing for the same amount of CPU. It is not like there would be a worker thread starvation issue.

**Comment by Ambrevar** [comment]
> You cannot run multiple pmap concurrently.
>
> Consider:
>
> ```
> (pmap ...)
> (pmap ...)
> ```

Misunderstanding, I meant

```
(bt:make-thread (lambda () (pmap ...)))
(bt:make-thread (lambda () (pmap ...)))
```

**Comment by Ambrevar** [comment]
I could be slower because pmap may induce more context switching.  I
don't know the details of lparallel, and I suppose we would need to measure
to say anything significant here.

> I say it is no slower than a normal mapcar because in both cases two mapcars running on two threads will also be competing for the same amount of CPU. It is not like there would be a worker thread starvation issue.

There can be, why not?
Actually, I've just opened an issue about it:

https://github.com/atlas-engineer/nyxt/issues/1684

**Comment by jmercouris** [comment]
Unless the amount of threads is greater than the amount of lparallel workers, there cannot be a worker starvation issue.

**Comment by jmercouris** [comment]
Further, this is a tunable during runtime and should either be set by the user or automatically determined by the system by running some internal benchmarks/profiling.

**Comment by Ambrevar** [comment]
What's the difference between a thread and a lparallel worker?

**Comment by Ambrevar** [comment]
> Further, this is a tunable during runtime and should either be set by the user or automatically determined by the system by running some internal benchmarks/profiling.

This is one of the reasons why the lparallel model is not very
apprioriate for Nyxt: we can't know the target systems in advance.

**Comment by mdbergmann** [comment]
I don't want to interrupt the discussion.
For performance it certainly depends on what exactly is computed with a mapcar/pmap.
But due to the overhead of splitting up a list into chunks, feeding then to a parallel map operation and later combining the results this makes only sense if the map function does significant computations.

**Comment by jmercouris** [comment]
This has been largely addressed. Though we could still parallelize more things :-)
